src/beamlime/constructors/providers.py

```diff
@@ -208,70 +212,204 @@ def __copy__(cls, _obj: Provider[Product]) -> Provider[Product]:
+            and self._all_kwargs_same_instances(kwargs)
+        )
+
+    def save_arguments(self, args: tuple, kwargs: dict) -> None:
+        self.args = args
+        self.kwargs = kwargs
 
+    def was_called(self) -> bool:
+        return hasattr(self, 'args') and hasattr(self, 'kwargs')
+
+    def __call__(self, *args: Any, **kwargs: Any) -> bool:
+        if not self.was_called():
+            self.save_arguments(args, kwargs)
+            return True
+        else:
+            return self.filter_arguments(args, kwargs)
+
+
+class _ArgumentsHashFilter(_ArgumentsInstanceFilter):
+    """Filter arguments by hash keys."""
 
-class CachedArguments:
     def __init__(self) -> None:
-        self.args: tuple[Any, ...]
-        self.kwargs: dict[str, Any]
+        super().__init__()
+        self.indicator = lru_cache(maxsize=2)(lambda *args, **kwargs: None)
+
+    def save_arguments(self, args: tuple, kwargs: dict) -> None:
+        self.indicator(*args, **kwargs)
+        return super().save_arguments(args, kwargs)
+
+    def filter_arguments(self, args: tuple, kwargs: dict) -> bool:
+        self.indicator(*args, **kwargs)
+        if self.indicator.cache_info().currsize > 1:
+            self.indicator.cache_clear()
+            self.indicator(*args, **kwargs)  # Reset cache info.
+            return False
+        return True
+
+
+_Item = TypeVar("_Item")
+
+
+def split_sequence_by_filter(
+    filter_func: Callable[[_Item], bool], sequence: Iterable[_Item]
+) -> Tuple[Tuple[_Item, ...], Tuple[_Item, ...]]:
+    """Split a sequence into two sequences based on the filter.
+
+    The first one of returned sequences contains items that pass the filter.
+    The other sequence contains the rest.
+    """
+    reference = tuple(filter(filter_func, sequence))
+    filtered = tuple(filter(lambda x: x not in reference, sequence))
+    return reference, filtered
+
+
+_Key = TypeVar("_Key")
+_Value = TypeVar("_Value")
+
+
+def split_dict_by_filter(
+    filter_func: Callable[[_Key, _Value], bool], sequence: dict[_Key, _Value]
+) -> Tuple[dict[_Key, _Value], dict[_Key, _Value]]:
+    """Split a dictionary into two dictionaries based on the filter.
+
+    The first one of returned dictionaries contains items that pass the filter.
+    The other dictionary contains the rest.
+    """
+    reference = {
+        key: value for key, value in sequence.items() if filter_func(key, value)
+    }
+    filtered = {key: value for key, value in sequence.items() if key not in reference}
+    return reference, filtered
+
+
+def _is_hashable(obj: object) -> bool:
+    return obj.__hash__ is not None
+
+
+class ArgumentsFilter(_ArgumentsInstanceFilter):
+    """
+    Remember arguments and check if the new arguments match the first one.
+    """
+
+    def __init__(self, func: Callable[..., Any]) -> None:
+        from inspect import Signature, signature
+
+        self._sig: Signature = signature(func)
+        self.unhashable_filter = _ArgumentsInstanceFilter()
```

Comment:
I find it a little strange that the `ArgumentsFilter` both inherits from the `_ArgumentsInstanceFilter` and has another instance of `_ArgumentsInstanceFilter` as a member... Does it really need to inherit as well?

Reply:
Yes, because if all `arguments` are the same instances, it does not have to split them into two groups (hashable/unhashable). Therefore `ArgumentsFilter` could use the `_ArgumentsInstanceFilter` methods.

I could just have 3 members,
1. `_ArgumentsInstanceFilter` that checks all `arguments`
2. `_ArgumentsInstanceFilter` that only checks unhashable `arguments`
3. `_ArgumentHashFilter` that only checks hashable `arguments`.

I also thought about only allowing `singleton` dependencies as JL suggested, but then it seems like too strict... that'll end up wrapping all providers with `SingletonProvider`...

Reply:
Yeah that sounds like a very good idea....!